[exporterhelper] Fix potential deadlocks in BatcherSender shutdown

dmitryax opened the pull request:
Fixes https://github.com/open-telemetry/opentelemetry-collector/issues/10255

TylerHelmuth commented:
I'm guessing I missed something since `TestBatchSender_MergeOrSplit` isn't passing anymore

dmitryax commented:
`TestBatchSender_MergeOrSplit` is good. It's just an accidental debug output, removed https://github.com/open-telemetry/opentelemetry-collector/pull/10261

The actual regression test was still failing. I updated the code, should be good now.

crobert-1 commented on exporter/exporterhelper/batch_sender_test.go:
```diff
@@ -439,6 +439,63 @@ func TestBatchSender_WithBatcherOption(t *testing.T) {
+	for i := 0; i < 10; i++ {
+		startWG.Add(1)
+		go func() {
+			startWG.Done()
```
```suggestion
			defer startWG.Done()
```
This should be after the `be.send` call, right? Maybe use defer instead? Or are you purposefully wanting to continue in the main goroutine before send is complete?

dmitryax replied:
No, it should be before send. We need to ensure all the goroutines are started. The send is blocking

crobert-1 replied:
> We need to ensure all the goroutines are started. The send is blocking

The goroutines will all be started regardless of where the `startWG.Done()` call is made, if I'm following correctly. The only impact is when the main goroutine will continue after the `startWG.Wait()` call. I'm wondering if it should wait until all sends are complete or not?

dmitryax replied:
the send operations cannot complete until they are unblocked, which we do after the shutdown - that's the purpose of the test case

crobert-1 replied:
Okay, thanks for clarifying 👍

crobert-1 commented on exporter/exporterhelper/batch_sender_test.go:
```diff
@@ -439,6 +439,63 @@ func TestBatchSender_WithBatcherOption(t *testing.T) {
+
+	// The exporter should have sent only one "merged" batch, in some cases it might send two if the shutdown
+	// happens before the batch is fully merged.
+	assert.LessOrEqual(t, uint64(1), sink.requestsCount.Load())
```
Is it valid to send two batches in some cases, as your comment is referencing? If so, should the assert here be `2` instead of `1`?

dmitryax replied:
It's usually 1 or more in rare cases. LessOrEqual checks that the first argument is less or equal to the second one
